app/code/core/Mage/Usa/Model/Shipping/Carrier/Ups.php

```diff
@@ -1584,13 +1455,403 @@ protected function _sendShipmentAcceptRequest(SimpleXMLElement $shipmentConfirmR
+
+        $this->_debug($debugData);
+        return $result;
+    }
+
+    protected function _doShipmentRequest(Varien_Object $request): Varien_Object
+    {
+        // this "if" will be removed after XML APIs will be shut down
+        if ($this->getConfigData('type') == 'UPS_XML') {
+            return $this->_doShipmentRequestXML($request);
+        }
+
+        // REST is default
+        return $this->_doShipmentRequestRest($request);
+    }
+
     /**
      * Do shipment request to carrier web service, obtain Print Shipping Labels and process errors in response
+     */
+    protected function _doShipmentRequestRest(Varien_Object $request): Varien_Object
+    {
+        $request->setShipperAddressCountryCode(
+            $this->getNormalizedCountryCode(
+                $request->getShipperAddressCountryCode(),
+                $request->getShipperAddressStateOrProvinceCode(),
+                $request->getShipperAddressPostalCode(),
+            )
+        );
+
+        $request->setRecipientAddressCountryCode(
+            $this->getNormalizedCountryCode(
+                $request->getRecipientAddressCountryCode(),
+                $request->getRecipientAddressStateOrProvinceCode(),
+                $request->getRecipientAddressPostalCode(),
+            )
+        );
+
+        $result = new Varien_Object();
+        $this->_prepareShipmentRequest($request);
+        $rawJsonRequest = $this->_formShipmentRestRequest($request);
+        $accessToken = $this->setAPIAccessRequest();
+        $this->_debug(['request_quote' => $rawJsonRequest]);
+
+        if ($this->getConfigFlag('mode_xml')) {
+            $shipConfirmUrl = $this->_liveUrls['ShipRestConfirm'];
+        } else {
+            $shipConfirmUrl = $this->_defaultUrls['ShipRestConfirm'];
+        }
+
+        /** Rest API Payload */
+        $headers = [
+            "Authorization: Bearer {$accessToken}",
+            "Content-Type: application/json"
+        ];
+        $debugData = [
+            'request' => $rawJsonRequest
+        ];
+
+        try {
+            $ch = curl_init();
+            curl_setopt($ch, CURLOPT_URL, $shipConfirmUrl);
+            curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+            curl_setopt($ch, CURLOPT_HEADER, 0);
+            curl_setopt($ch, CURLOPT_POST, 1);
+            curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
+            curl_setopt($ch, CURLOPT_POSTFIELDS, $rawJsonRequest);
+            curl_setopt($ch, CURLOPT_TIMEOUT, 30);
+            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, $this->getConfigFlag('verify_peer'));
+            $responseData = curl_exec($ch);
+            curl_close($ch);
+        } catch (Exception $e) {
+            $debugData['result'] = ['error' => $e->getMessage(), 'code' => $e->getCode()];
+            Mage::logException($e);
+        }
+
+        if (!isset($responseData)) {
+            $responseData = '';
+        }
+        $responseData = json_decode($responseData, JSON_THROW_ON_ERROR);
```

Comment:
This call is incorrect. The flags are the fourth argument. This call should be `json_decode($responseData, false, 512, JSON_THROW_ON_ERROR);`.

It should also be wrapping in a try-catch if we don't want this method to be able to actually hard-fail.

Reply:
I've removed the JSON_THROW_ON_ERROR at the moment, thanks!